File: src/main/GO/leetcode/lists/offer/Offer36.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/4 15:15
 */

func treeToDoublyList(root *TreeNode) *TreeNode {
	var dfs func(cur *TreeNode)
	var pre, head *TreeNode
	dfs = func(cur *TreeNode) {
		if cur == nil {
			return
		}
		dfs(cur.Left)
		if pre != nil {
			pre.Right = cur
		} else {
			head = cur
		}
		cur.Left = pre
		pre = cur
		dfs(cur.Right)
	}
	dfs(root)
	return head
}

File: java/src/leetcode/lists/offer/Offer36.java
package leetcode.lists.offer;

import common.Node;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/9/12 23:36
 */
public class Offer36 {

    Node pre, head;

    public Node treeToDoublyList(Node root) {
        if (root == null) {
            return null;
        }
        dfs(root);
        head.left = pre;
        pre.right = head;
        return head;
    }

    private void dfs(Node root) {
        if (root == null) {
            return;
        }
        dfs(root.left);
        // do something
        if (pre != null) {
            pre.right = root;
        } else {
            head = root;
        }
        root.left = pre;
        pre = root;

        dfs(root.right);
    }
}
